test/functional/cursor_tests.js

```diff
@@ -4163,6 +4163,95 @@ describe('Cursor', function() {
     }
   });
 
+  it('Should properly kill a cursor', {
+    metadata: {
+      requires: {
+        topology: ['single', 'replicaset', 'sharded', 'ssl', 'heap', 'wiredtiger'],
+        mongodb: '>=3.2.0'
+      }
+    },
+
+    // The actual test we wish to run
+    test: function() {
+      // Load up the documents
+      const docs = [];
+      for (let i = 0; i < 1000; i += 1) {
+        docs.push({
+          a: i
+        });
+      }
+
+      const configuration = this.configuration;
+      const client = configuration.newClient(configuration.writeConcernMax(), { poolSize: 1 });
+      let cleanup = () => {};
+      let caughtError = undefined;
+
+      return (
+        client
+          // Connect
+          .connect()
+          .then(function(client) {
+            cleanup = () => client.close();
+            const db = client.db(configuration.db);
+            const collection = db.collection('cursorkilltest1');
+
+            // Insert 1000 documents
+            return collection.insert(docs).then(() => {
+              // Generate cursor for find operation
+              const cursor = collection.find({});
+
+              // Iterate cursor past first element
+              return cursor
```

Comment:
I get keeping the promise chain one level deeper so you can maintain access to the `client` variable, but it looks like right about here we can keep the chain just at the level of the `collection.insert(docs)` above - starting to look a little callback hellish :)

Reply:
oh I take it back, you're using the `cursor` as well, disregard